### ISSUE TITLE
Unable to process multiple preparations refering to the same sequencing run

### ISSUE THREAD

**Issue by wasade** [qiita-spots/qp-knight-lab-processing#51]
This refers to jobs:

- 490c4831-33c6-44a2-8a34-ddcf08d0c40d
- a78aa38d-5b82-4137-9950-8524414c4b51
- b9b6b094-3345-4808-80dc-7a4488abdb4f

In this situation, replicate samples were sequenced on the same lane. To account for this, and to ensure unique samples exist in the preparation, each set of replicates had its own sample sheet. Processing on the three sample sheets occurred in parallel. What appears to have happened is that the processing jobs all attempted to move the same files (relative to uploads/) to different locations at preparation creation.

It is important to note that the Qiita processing pipeline remarked these jobs as successful, even though the constructed preparations were unsuccessful.

**Comment by charles-cowart** [comment]
From another perspective, three QPP jobs were submitted using the same run-id (230411_M05314_0358_000000000-KYF72) and lane number (1), w/mapping-files containing an identical set of sample-names but with differing values for barcode, primer_plate, well_id, sample_plate, project_name, orig_name, and well_description. 

The following three sets of items get moved into the uploads directory:
A 'BLANKs' file: _230411_M05314_0358_000000000-KYF72_Ornish_PMRI_FOBT_Rep_3_10317_blanks.tsv_
the final set of fastq files: _QCJob/Ornish_PMRI_FOBT_Rep_3_10317/amplicon/*_
One or more prep-information files, depending on the number of projects in the mapping-file/sample-sheet: _230411_M05314_0358_000000000-KYF72.Ornish_PMRI_FOBT_Rep_3_10317.1.tsv_

In this case, the blanks and prep-information filenames are unique across the three jobs (Rep_1, Rep_2, Rep_3), so they shouldn't overwrite each other. However as their filenames are built using metadata supplied by the user, this is not always true.

The three jobs were also ingesting their set of prep-information files and updating the set of BLANKs in Qiita within the same time frame.

**Comment by charles-cowart** [comment]
(FYI) Our use-case for being able to start a QPP job on a run-id that was already processed was to overwrite previous results w/improved ones. This is still a valuable use-case and why overwriting data in the uploads directory was allowed originally. 

In the short term, running each job sequentially is the best option to avoid overwrites and unpredictable behavior. We end up processing the same BCL files more than once which is waste, but in this case the final fastq output isn't large by our standards (~4.7GB). It'll be faster to process them sequentially than to hack a short-term solution.

In the long term, submitting multiple jobs, one for each prep is generally wasteful as the two most resource-intensive steps in a QPP job (bcl->fastq conversion, adapter-trimming & human-filtering) are being performed once for each prep instead of sharing the same results. Instead we should make a small modification to the UI and the pipeline to accept multiple pre-prep files at the same time. QPP will then reuse the results of the most resource-intensive steps across all preps. We can introduce a lock on the run-id (alternatively the run-directory itself) and return an error if a user tries to submit a new job using a run-id that is currently being processed. This will prevent the unpredictable behavior from happening again.

There will be some slight changes to the layout of the final output folder to accommodate multiple preps and the copy commands in addition to some code changes to perform multiple FastQC and MultiQC operations, one against each prep.

**Comment by wasade** [comment]
Thanks @charles-cowart, this seems reasonable. And agree some notion of a lock would be valuable if the output filepaths are not assured to be unique to a processing job.

**Comment by wasade** [comment]
Hi Charlie, are there any updates on this issue?

**Comment by charles-cowart** [comment]
@wasade No updates on this issue. I'd like to run them by @antgonza first.

**Comment by charles-cowart** [comment]
Addressed by https://github.com/biocore/mg-scripts/pull/104

**Comment by charles-cowart** [comment]
run_dir locking is likely not needed since SPP no longer directly moves files into Qiita's uploads directory. However I am still looking into being able to submit multiple pre-prep files to a pipeline in order to make use of the same bcl->fastq and potential QC steps.

**Comment by charles-cowart** [comment]
Discussed with Rodolfo and Antonio and we agreed that the only time submitting multiple preps would be needed is to support replicates. Since plate-level replication has now been implemented (but not yet deployed to production), multiple submissions are no longer needed and this issue can be closed.
